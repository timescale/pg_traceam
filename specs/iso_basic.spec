setup
{
    DROP TABLE IF EXISTS iso_basic;
    CREATE TABLE iso_basic USING traceam
              AS SELECT i FROM generate_series(1, 5) i;
}

teardown
{
    DROP TABLE iso_basic;
}

session s1
step s1b1 { BEGIN ISOLATION LEVEL READ COMMITTED; }
step s1b2 { BEGIN ISOLATION LEVEL REPEATABLE READ; }
step s1u { UPDATE iso_basic SET i = i * 3; }
step s1c { COMMIT; }
teardown { SELECT * FROM iso_basic; }

session s2
step s2b1 { BEGIN ISOLATION LEVEL READ COMMITTED; }
step s2b2 { BEGIN ISOLATION LEVEL REPEATABLE READ; }
step s2u { UPDATE iso_basic SET i = i * 2; }
step s2c { COMMIT; }
teardown { SELECT * FROM iso_basic; }

# This is probably overkill; some of these permutations may not contain
# additional useful coverage.
permutation s1b1 s2b1 s1u s2u s1c s2c
permutation s1b1 s2b1 s1u s1c s2u s2c
permutation s1b2 s2b2 s1u s2u s1c s2c
permutation s1b2 s2b2 s1u s1c s2u s2c
permutation s1b1 s2b2 s1u s2u s1c s2c
permutation s1b1 s2b2 s1u s1c s2u s2c
permutation s1b2 s2b1 s1u s2u s1c s2c
permutation s1b2 s2b1 s1u s1c s2u s2c
