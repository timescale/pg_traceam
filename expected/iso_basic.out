Parsed test spec with 2 sessions

starting permutation: s1b1 s2b1 s1u s2u s1c s2c
step s1b1: BEGIN ISOLATION LEVEL READ COMMITTED;
step s2b1: BEGIN ISOLATION LEVEL READ COMMITTED;
step s1u: UPDATE iso_basic SET i = i * 3;
step s2u: UPDATE iso_basic SET i = i * 2; <waiting ...>
step s1c: COMMIT;
step s2u: <... completed>
step s2c: COMMIT;
 i
--
 6
12
18
24
30
(5 rows)

 i
--
 6
12
18
24
30
(5 rows)


starting permutation: s1b1 s2b1 s1u s1c s2u s2c
step s1b1: BEGIN ISOLATION LEVEL READ COMMITTED;
step s2b1: BEGIN ISOLATION LEVEL READ COMMITTED;
step s1u: UPDATE iso_basic SET i = i * 3;
step s1c: COMMIT;
step s2u: UPDATE iso_basic SET i = i * 2;
step s2c: COMMIT;
 i
--
 6
12
18
24
30
(5 rows)

 i
--
 6
12
18
24
30
(5 rows)


starting permutation: s1b2 s2b2 s1u s2u s1c s2c
step s1b2: BEGIN ISOLATION LEVEL REPEATABLE READ;
step s2b2: BEGIN ISOLATION LEVEL REPEATABLE READ;
step s1u: UPDATE iso_basic SET i = i * 3;
step s2u: UPDATE iso_basic SET i = i * 2; <waiting ...>
step s1c: COMMIT;
step s2u: <... completed>
ERROR:  could not serialize access due to concurrent update
step s2c: COMMIT;
 i
--
 3
 6
 9
12
15
(5 rows)

 i
--
 3
 6
 9
12
15
(5 rows)


starting permutation: s1b2 s2b2 s1u s1c s2u s2c
step s1b2: BEGIN ISOLATION LEVEL REPEATABLE READ;
step s2b2: BEGIN ISOLATION LEVEL REPEATABLE READ;
step s1u: UPDATE iso_basic SET i = i * 3;
step s1c: COMMIT;
step s2u: UPDATE iso_basic SET i = i * 2;
step s2c: COMMIT;
 i
--
 6
12
18
24
30
(5 rows)

 i
--
 6
12
18
24
30
(5 rows)


starting permutation: s1b1 s2b2 s1u s2u s1c s2c
step s1b1: BEGIN ISOLATION LEVEL READ COMMITTED;
step s2b2: BEGIN ISOLATION LEVEL REPEATABLE READ;
step s1u: UPDATE iso_basic SET i = i * 3;
step s2u: UPDATE iso_basic SET i = i * 2; <waiting ...>
step s1c: COMMIT;
step s2u: <... completed>
ERROR:  could not serialize access due to concurrent update
step s2c: COMMIT;
 i
--
 3
 6
 9
12
15
(5 rows)

 i
--
 3
 6
 9
12
15
(5 rows)


starting permutation: s1b1 s2b2 s1u s1c s2u s2c
step s1b1: BEGIN ISOLATION LEVEL READ COMMITTED;
step s2b2: BEGIN ISOLATION LEVEL REPEATABLE READ;
step s1u: UPDATE iso_basic SET i = i * 3;
step s1c: COMMIT;
step s2u: UPDATE iso_basic SET i = i * 2;
step s2c: COMMIT;
 i
--
 6
12
18
24
30
(5 rows)

 i
--
 6
12
18
24
30
(5 rows)


starting permutation: s1b2 s2b1 s1u s2u s1c s2c
step s1b2: BEGIN ISOLATION LEVEL REPEATABLE READ;
step s2b1: BEGIN ISOLATION LEVEL READ COMMITTED;
step s1u: UPDATE iso_basic SET i = i * 3;
step s2u: UPDATE iso_basic SET i = i * 2; <waiting ...>
step s1c: COMMIT;
step s2u: <... completed>
step s2c: COMMIT;
 i
--
 6
12
18
24
30
(5 rows)

 i
--
 6
12
18
24
30
(5 rows)


starting permutation: s1b2 s2b1 s1u s1c s2u s2c
step s1b2: BEGIN ISOLATION LEVEL REPEATABLE READ;
step s2b1: BEGIN ISOLATION LEVEL READ COMMITTED;
step s1u: UPDATE iso_basic SET i = i * 3;
step s1c: COMMIT;
step s2u: UPDATE iso_basic SET i = i * 2;
step s2c: COMMIT;
 i
--
 6
12
18
24
30
(5 rows)

 i
--
 6
12
18
24
30
(5 rows)

